src/configuration.c

```diff
@@ -0,0 +1,202 @@
+#include <stdlib.h>
+#ifdef WITH_LIBCONFIG
+#include <string.h>
+#include <ctype.h>
+#include <stdbool.h>
+#include <libconfig.h>
+#include <basedir_fs.h>
+#endif
+
+#include "configuration.h"
+#include "memory.h"
+
+#define KEY_CTRL(key) ((const char[]){((key) - ('@')), '\0'})
+
+typedef struct {
+	const char *key;
+	void (*action)(tty_interface_t *);
+} indexed_action_t;
+
+indexed_action_t default_keybindings[] = {
+	{"\x7f", action_del_char},	/* DEL */
+	{KEY_CTRL('H'), action_del_char}, /* Backspace (C-H) */
+	{KEY_CTRL('W'), action_del_word}, /* C-W */
+	{KEY_CTRL('U'), action_del_all},  /* C-U */
+	{KEY_CTRL('I'), action_autocomplete}, /* TAB (C-I ) */
+	{KEY_CTRL('C'), action_exit},	 /* C-C */
+	{KEY_CTRL('D'), action_exit},	 /* C-D */
+	{KEY_CTRL('M'), action_emit},	 /* CR */
+	{KEY_CTRL('P'), action_prev},	 /* C-P */
+	{KEY_CTRL('N'), action_next},	 /* C-N */
+	{KEY_CTRL('K'), action_prev},	 /* C-J */
+	{KEY_CTRL('J'), action_next},	 /* C-K */
+	{"\x1b[A", action_prev}, /* UP */
+	{"\x1bOA", action_prev}, /* UP */
+	{"\x1b[B", action_next}, /* DOWN */
+	{"\x1bOB", action_next}, /* DOWN */
+	{"\x1b[5~", action_pageup},
+	{"\x1b[6~", action_pagedown}
+};
+
+#ifdef WITH_LIBCONFIG
+indexed_action_t actions[] = {
+	{"emit", action_emit},
+	{"del_char", action_del_char},
+	{"del_word", action_del_word},
+	{"del_all", action_del_all},
+	{"prev", action_prev},
+	{"next", action_next},
+	{"pageup", action_pageup},
+	{"pagedown", action_pagedown},
+	{"autocomplete", action_autocomplete},
+	{"exit", action_exit}
+};
+
+action_t action_by_name(char const *name) {
+	const size_t action_count = sizeof(actions) / sizeof(indexed_action_t);
+	for (size_t i = 0; i < action_count; ++i) {
+		if (! strcmp(actions[i].key, name)) {
+			return actions[i].action;
+		}
+	}
+
+	fprintf(stderr, "action named `%s' does not exist\n", name);
+	exit(1);
+}
+
+static bool insert_keybinding(keybinding_t *keybindings, size_t keybinding_count, const char *key, action_t action) {
+	for (size_t i = 0; i < keybinding_count; ++i) {
+		if (! strcmp(keybindings[i].key, key)) {
+			keybindings[i].action = action;
+			return false;
+		}
+	}
+
+	keybindings[keybinding_count].key = safe_strdup(key);
```

Comment:
We know there is enough space to accommodate the keybinding here from the realloc on line 166.